can/bus.py

```diff
@@ -2,79 +2,174 @@
 """
 
 from __future__ import print_function, absolute_import
 
 from abc import ABCMeta, abstractmethod
 import logging
 import threading
+from time import time
 from collections import namedtuple
 
-from can.broadcastmanager import ThreadBasedCyclicSendTask
-
-logger = logging.getLogger(__name__)
+from .broadcastmanager import ThreadBasedCyclicSendTask
 
 
 BusState = namedtuple('BusState', 'ACTIVE, PASSIVE, ERROR')
 
 
 class BusABC(object):
-    """CAN Bus Abstract Base Class
-
-    Concrete implementations must implement the following methods:
-        * send
-        * recv
-
-    As well as setting the `channel_info` attribute to a string describing the
-    interface.
-
-    They may implement :meth:`~can.BusABC._detect_available_configs` to allow
-    the interface to report which configurations are currently available for
-    new connections.
+    """The CAN Bus Abstract Base Class that serves as the basis
+    for all concrete interfaces.
+
+    Concrete implementations *have to* implement the following:
+        * :meth:`~can.BusABC.send` to send individual messages
+        * :meth:`~can.BusABC._recv_internal` to receive individual messages
+          (see note below)
+        * set the :attr:`~can.BusABC.channel_info` attribute to a string describing
+          the underlying bus and/or channel
+
+    They *might* implement the following:
+        * :meth:`~can.BusABC.flush_tx_buffer` to allow discrading any
+          messages yet to be sent
+        * :meth:`~can.BusABC.shutdown` to override how the bus should
+          shut down
+        * :meth:`~can.BusABC.send_periodic` to override the software based
+          periodic sending and push it down to the kernel or hardware
+        * :meth:`~can.BusABC._apply_filters` to apply efficient filters
+          to lower level systems like the OS kernel or hardware
+        * :meth:`~can.BusABC._detect_available_configs` to allow the interface
+          to report which configurations are currently available for new
+          connections
+        * :meth:`~can.BusABC.state` property to allow reading and/or changing
+          the bus state
+
+    .. note::
+
+       Previously, concrete bus classes had to override :meth:`~can.BusABC.recv`
+       directly instead of :meth:`~can.BusABC._recv_internal`, but that has
+       changed to allow the abstract base class to handle in-software message
+       filtering as a fallback. Older (custom) interfaces might still be
+       implemented like that and thus might not provide message filtering.
 
     """
 
-    #: a string describing the underlying bus channel
+    #: a string describing the underlying bus and/or channel
     channel_info = 'unknown'
 
     @abstractmethod
     def __init__(self, channel=None, can_filters=None, **config):
-        """
+        """Construct and open a CAN bus instance of the specified type.
+
+        Subclasses should call though this one with all parameters as
+        it applies filters (for now).
```

Comment:
Keep in mind that a lot of the code comments have lasted many many years so when people read "for now" they won't easily have any idea if that is talking of change to come soon... or was an idea the developer had 5 years earlier...